packages/datadog-plugin-undici/src/index.js

```diff
@@ -0,0 +1,341 @@
+  if (!options) {
+    return false
+  }
+
+  if (options.headers) {
+    let headers = {}
+    if (typeof options.headers === 'string') {
+      headers = parseHeaders(options.headers)
+    } else {
+      headers = Object.keys(options.headers).reduce(
+        (prev, next) =>
+          Object.assign(prev, {
+            [next.toLowerCase()]: options.headers[next]
+          }),
+        {}
+      )
+    }
+
+    if (headers['x-amz-signature']) {
+      return true
+    }
+
+    if (
+      [].concat(headers['authorization']).some(startsWith('AWS4-HMAC-SHA256'))
+    ) {
+      return true
+    }
+  }
+
+  return (
+    options.path &&
+    options.path.toLowerCase().indexOf('x-amz-signature=') !== -1
+  )
+}
+
+function startsWith (searchString) {
+  return (value) => String(value).startsWith(searchString)
+}
+
+function getStatusValidator (config) {
+  if (typeof config.validateStatus === 'function') {
+    return config.validateStatus
+  } else if (config.hasOwnProperty('validateStatus')) {
+    log.error('Expected `validateStatus` to be a function.')
+  }
+  return (code) => code < 400 || code >= 500
+}
+
+function getFilter (tracer, config) {
+  const blocklist = tracer._url ? [getAgentFilter(tracer._url)] : []
+
+  config = Object.assign({}, config, {
+    blocklist: blocklist.concat(config.blocklist || [])
+  })
+
+  return urlFilter.getFilter(config)
+}
+
+function getAgentFilter (url) {
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
+  const agentFilter = url.href.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
+
+  return RegExp(`^${agentFilter}.*$`, 'i')
+}
+
+function normalizeConfig (tracer, config) {
+  const validateStatus = getStatusValidator(config)
+  const filter = getFilter(tracer, config)
+  const propagationFilter = getFilter(tracer, {
+    blocklist: config.propagationBlocklist
+  })
+  const headers = getHeaders(config)
+  const hooks = getHooks(config)
+
+  return Object.assign({}, config, {
+    validateStatus,
+    filter,
+    propagationFilter,
+    headers,
+    hooks
+  })
+}
+
+function getHeaders (config) {
+  if (!Array.isArray(config.headers)) return []
+
+  return config.headers
+    .filter((key) => typeof key === 'string')
+    .map((key) => key.toLowerCase())
+}
+
+function getHooks (config) {
+  const noop = () => {}
+  const request = (config.hooks && config.hooks.request) || noop
+
+  return { request }
+}
+
+function patch (undici, methodName, tracer, config) {
+  this.wrap(undici, methodName, fn => makeRequestTrace(fn))
+
+  function makeRequestTrace (request) {
+    return function requestTrace () {
+      // Bind the callback for async resources
+      if (arguments.length === 3) {
+        arguments[2] = AsyncResource.bind(arguments[2])
+      }
+      const scope = tracer.scope()
+      const childOf = scope.active()
+      const span = tracer.startSpan('http.request', {
+        childOf,
+        tags: {
+          [SPAN_KIND]: CLIENT
+        }
+      })
```

Comment:
we init the span early and pass it through the asyncLocalStorage. not sure this is exactly what was intended